src/nn/conv.rs

```diff
@@ -0,0 +1,256 @@
+            BATCH_SIZE,
+            IN_CHAN,
+            OUT_CHAN,
+            KERNEL_SIZE,
+            STRIDE,
+            PADDING,
+            IN_HEIGHT,
+            IN_WIDTH,
+        >(x, &self.weight, &self.bias)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::{prelude::StdRng, SeedableRng};
+    use std::fs::File;
+    use tempfile::NamedTempFile;
+
+    #[test]
+    fn test_forward_3d_sizes() {
+        type Img = Tensor3D<3, 10, 10>;
+        let _: Tensor3D<2, 8, 8> = Conv2D::<3, 2, 3>::default().forward(Img::zeros());
+        let _: Tensor3D<4, 8, 8> = Conv2D::<3, 4, 3>::default().forward(Img::zeros());
+        let _: Tensor3D<4, 9, 9> = Conv2D::<3, 4, 2>::default().forward(Img::zeros());
+        let _: Tensor3D<4, 7, 7> = Conv2D::<3, 4, 4>::default().forward(Img::zeros());
+        let _: Tensor3D<2, 4, 4> = Conv2D::<3, 2, 3, 2>::default().forward(Img::zeros());
+        let _: Tensor3D<2, 3, 3> = Conv2D::<3, 2, 3, 3>::default().forward(Img::zeros());
+        let _: Tensor3D<2, 10, 10> = Conv2D::<3, 2, 3, 1, 1>::default().forward(Img::zeros());
+        let _: Tensor3D<2, 12, 12> = Conv2D::<3, 2, 3, 1, 2>::default().forward(Img::zeros());
+        let _: Tensor3D<2, 6, 6> = Conv2D::<3, 2, 3, 2, 2>::default().forward(Img::zeros());
+    }
+
+    #[test]
+    fn test_forward_4d_sizes() {
+        type Img = Tensor4D<5, 3, 10, 10>;
+        let _: Tensor4D<5, 2, 8, 8> = Conv2D::<3, 2, 3>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 4, 8, 8> = Conv2D::<3, 4, 3>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 4, 9, 9> = Conv2D::<3, 4, 2>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 4, 7, 7> = Conv2D::<3, 4, 4>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 2, 4, 4> = Conv2D::<3, 2, 3, 2>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 2, 3, 3> = Conv2D::<3, 2, 3, 3>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 2, 10, 10> = Conv2D::<3, 2, 3, 1, 1>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 2, 12, 12> = Conv2D::<3, 2, 3, 1, 2>::default().forward(Img::zeros());
+        let _: Tensor4D<5, 2, 6, 6> = Conv2D::<3, 2, 3, 2, 2>::default().forward(Img::zeros());
+    }
+
+    #[test]
+    fn test_2_conv_sizes() {
+        type A = Conv2D<1, 2, 3>;
+        type B = Conv2D<2, 4, 3>;
+        let _: Tensor3D<4, 6, 6> = B::default().forward(Tensor3D::<2, 8, 8>::zeros());
+        let _: Tensor3D<4, 6, 6> = <(A, B)>::default().forward(Tensor3D::<1, 10, 10>::zeros());
+    }
+
+    #[test]
+    fn test_3_conv_sizes() {
+        type A = Conv2D<1, 2, 3>;
+        type B = Conv2D<2, 4, 3>;
+        type C = Conv2D<4, 1, 1, 1, 1>;
+
+        type Img = Tensor3D<1, 10, 10>;
+        let _: Tensor3D<2, 8, 8> = A::default().forward(Img::zeros());
+        let _: Tensor3D<4, 6, 6> = <(A, B)>::default().forward(Img::zeros());
+        let _: Tensor3D<1, 8, 8> = C::default().forward(Tensor3D::<4, 6, 6>::zeros());
```

Comment:
So something with generic const exprs is currently broken. If you remove this line you will get a bunch of linker errors about symbols not being included. This makes convs pretty unusable at this point 😞 .

Reply:
Yup, I could reproduce this as well. But even apart from the extern symbols there is also something weird going on with the module.

```rust
  = note: /usr/bin/ld: /home/void/Projects/Rust/dfdx/target/debug/deps/dfdx-5aab129f28193da6.5gs3y0we70clq0q0.rcgu.o: in function `dfdx::nn::impl_module_for_tuples::<impl dfdx::nn::module::Module<Input> for (A,B,C)>::forward':
          /home/void/Projects/Rust/dfdx/src/nn/impl_module_for_tuples.rs:79: undefined reference to `<dfdx::nn::conv::Conv2D<_,_,_,_,_> as dfdx::nn::module::Module<dfdx::tensor::structs::Tensor3D<_,_,_,TAPE>>>::forward'
          /usr/bin/ld: /home/void/Projects/Rust/dfdx/target/debug/deps/dfdx-5aab129f28193da6: hidden symbol `_ZN155_$LT$dfdx..nn..conv..Conv2D$LT$_$C$_$C$_$C$_$C$_$GT$$u20$as$u20$dfdx..nn..module..Module$LT$dfdx..tensor..structs..Tensor3D$LT$_$C$_$C$_$C$TAPE$GT$$GT$$GT$7forward17h4239bcee10941f68E' isn't defined
          /usr/bin/ld: final link failed: bad value
```

Reply:
This is fixed on the latest nightly